[2단계 - 자동차 경주 구현] 카고(정민호) 미션 제출합니다.

kargowild opened the pull request:
안녕하세요 비밥!
step1에서 머지 이후  getWinners 메서드에 대해 질문드렸던 부분 다시 언급드립니다!

비밥 : 
Referee는 상태를 가지지 않고 static한 정적 메서드만 가지고 있는것 같네요.
상태를 가지고 있는 객체를 생성하는 건 어떨까요?
일급컬렉션도 단일책임원칙을 가지는 것이 좋아보입니다.
우승자의 정보를 얻기 위해 winnerPosition을 구하기 위해 `Accelerator`가 필요할까요?
필요하지 않다면 응집도가 높다고 볼 수 있을까요?
여러가지 고민을 해보시고 다른방향으로 구현해보시는 것도 좋을것 같습니다 😄

카고:
다시 보니 `Cars`가 `Accelerator`를 멤버변수 가지고 있을 필요는 없어보여서 `tryMove`를 호출할 때 매개변수로 `Accelerator`를 받도록 변경함으로써 `Cars`가 `List<Car>`외에 다른 멤버변수를 갖지 않는 일급컬렉션으로 만들었습니다!

우선 `Cars`가 오직 `경주에 참여하는 자동차들을 동시에 움직이게 한다`라는 책임만 갖고 있다고 보았고, 따라서 `winnerPosition`을 구하는 책임은 `Referee`에게 이전해봤어요. 또한 `Referee`가 `Cars`에 의존할 필요는 없고 `자동차이름 : 위치` 에 해당하는 Map 정보만 알면 될 것 같아서 멤버변수로 `Cars`가 아닌 `Map<String, Integer> carsPosition`을 갖도록 설계해보았어요.

이렇게 장황하게 설명을 드린 이유는 한 가지 고민되는 포인트가 있기 때문입니다..!

프리코스를 진행하며 `객체가 가진 상태를 getter로 가져와 외부에서 처리하지 말고, 상태를 갖고 있는 객체에게 직접 시켜라`라는 글들을 여럿 보았는데, `Referee`가 현재 하는 일이 `Cars`의 상태를 getter로 가져와 외부에서 처리하는 듯한 느낌을 받았기 때문이에요. 비밥이 보기엔 어떤가요? 현재 역할 분담이 괜찮아 보이나요??

아래 커밋에 위 내용을 반영해보았는데, 현재 PR이 아닌 2단계 PR에 포함될 것 같아요. 2단계 리뷰때 확인 부탁드립니다!! 🙇‍♂️

refactor: winnerPosition을 구하는 역할을 Cars에서 Referee로 이전

kargowild committed:
refactor: CarRacingTest 파일의 디렉토리 위치를 적절하게 변경

kargowild committed:
refactor: CarRacing의 메서드 중 외부노출이 필요없는 메서드들을 private하게 숨김으로써 필요없어진 테스트 파일제거

kargowild committed:
refactor: Cars가 일급컬렉션이 되도록 수정

- Cars가 Accelerator를 멤버변수로 갖지 않고 tryMove 메서드를 수행할 때 매개변수로 받게 만듦
- 따라서 Cars는 멤버변수로 오직 List<Car>만을 갖게되어 일급컬렉션이 됨

kargowild committed:
refactor: winnerPosition을 구하는 역할을 Cars에서 Referee로 이전

Cars는 경주에 참여하는 자동차들을 동시에 움직이는 책임만을 갖도록 하고, 우승자를 가리기 위한 책임들은 모두 Referee가 갖도록 리팩토링함

kargowild committed:
test: 하나의 테스트 메서드에서만 사용되는 객체들은 테스트 메서드 내에서 생성하도록 변경

kargowild committed:
test: 의미 없는 공백 제거

kargowild committed:
test: 경계값만 테스트하도록 수정

kargowild committed:
refactor: io -> view 로 패키지명 변경

kargowild committed:
refactor: String 클래스의 repeat 메서드를 활용해 코드의 가독성 향상

kargowild committed:
refactor: 메서드의 위치를 가독성 좋게 변경

kargowild committed:
refactor: 변할 가능성이 없는 변수들에 final 키워드 추가

pci2676 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -18,25 +17,21 @@ public void printMoveResultMessage() {
         for (Car car : cars) {
             String positionBar = createPositionBar(car.getPosition());
-            System.out.println(String.format(CAR_POSITION_MESSAGE, car.getName(), positionBar));
+            System.out.printf((CAR_POSITION_MESSAGE) + "%n", car.getName(), positionBar);
```
car는 내부 값이 변경될 수 있는 메서드를 가지고 있습니다.

view에서 이러한 메서드가 호출되지 않을거라는 확신을 가질 수 있을까요?

kargowild replied:
그렇네요. view에게 Car 객체를 직접 넘겨주는 것은 위험할 것 같아요.

OutputView는 `String carName, int position`을 받아 자동차 하나의 정보에 대해 출력할 수 있도록 하고, 여러 자동차들을 출력시키려면 컨트롤러에서 for문을 돌도록 리팩토링해보았습니다!

kargowild replied:
IO 호출 빈도에 대해 고민해본 후 다시 코드를 보니 기존에 Cars를 넘겨주던 방식은 StringBuilder를 사용하여 IO 호출 빈도를 낮추는 방향으로 개선할 수 있을 것 같아요. 

Car 객체를 view에 넘겨주지 않고, IO 호출 빈도도 낮추도록 설계하려면 당장 생각나는 방법은 view가 매개변수로 `List<String> carName, List<Integer> position` 또는 `Map<String, Integer> carsPosition`을 받아서 StringBuilder로 설계하는 방법인데, 이에 대해 어떻게 생각하시나요 ?!

pci2676 replied:
List<String> carName, List<Integer> position 는 두 컬렉션의 연관성이 떨어져 보일 수 있으니 Map<String, Integer> carsPosition의 형태가 더 좋을것 같습니다.

pci2676 commented on src/main/java/racingcar/CarRacing.java:
```diff
@@ -1,60 +1,60 @@
     }
+
+    private void printWinners(final Cars cars) {
+        final Referee referee = new Referee(cars.getCarsPosition());
```
이 부분에서 고민을 하신것 같습니다.

cars의 move 이후 반환되는 값이 있다면 이를 모아서 referee에게 전달하면 getter를 사용하지 않고 문제를 해결할 수 있지 않을까요?

kargowild replied:
현재 설계상으로는 `cars`가 모든 자동차들을 1번만 움직이도록 되어있어요. 조언해주신대로 `cars`의 `tryMove`이후 반환값을 만든다면, 아마 아래와 같이 할 것 같아요.

- `List<String>`으로 이동에 성공한 자동차들만 반환해주면, `Referee`는 `tryCount` 만큼 이동이 발생할 때마다 `List<String>` 값을 받아 이름이 매칭되는 자동차들의 Position을 1씩 늘려가며 `자동차이름 - 포지션` 에 해당하는 맵을 스스로 만들 수도 있을 것 같아요.

그런데 한편으로는, `Car` 객체가 이미 `자동차이름 - 포지션`을 가지고 있는 객체이고 `Cars`는 이런 `Car`들을 List로 가지고 있는데,`List<Car>`를 넘겨주는게 깔끔하지 않나? 하는 생각도 듭니다.

그리고 또 이렇게 생각하다보니, `Cars`가 이미 `List<Car>`를 가지고 있는데 `Cars`가 우승자를 가려내는게 자신이 가진 상태에 대해 적절하게 책임을 지는게 아닐까? 하는 의문도 들어요. 

어떤 선택이 좋은 방향일지 갈팡질팡합니다 ㅠㅠ 비밥의 의견은 어떤가요?

pci2676 replied:
`List<Car>`를 넘기는 것은 위험할 수 있습니다. printWinners에서 car의 move를 호출하지 않을거라는 보장을 할 수 없습니다. 읽기 전용 모델을 이용하여 전달하는게 더 안전할 것 같습니다.

Cars에서 우승자를 가려내는것은 Cars의 move에 영향을 받는것 같습니다. move에 영향을 받지 않고 항상 동일한 결과를 보장해주면 값을 추적하기 쉽지 않을까요? 그러한 관점에서 move의 메서드 응답값을 이용하면 불변한 값을 안전하게 얻을 수 있을것 같습니다.

kargowild replied:
그렇군요.
메서드의 반환값이 getter를 대체할 수 있다면 적극 사용하는 것이 좋고,
객체의 정보에 대해 다른 객체에게 전달해주어야 한다면 읽기 전용 모델을 제공하는 것이 유리한 것 같습니다.

안전한 코드를 만들기 위해 생각해봐야하는 좋은 포인트인 것 같아요. 다음 미션들에서 꼭 적용해보겠습니다 !!

pci2676 commented on src/main/java/racingcar/CarRacing.java:
```diff
@@ -1,60 +1,60 @@
-
-    public Cars createCars(List<String> carNames) {
-        List<Car> cars = new ArrayList<>();
+    private Cars createCars(final List<String> carNames) {
+        final List<Car> cars = new ArrayList<>();
```
Stream을 이용해서 표현해보시는것도 한번 쯤 시도해보시면 좋을것 같습니다.

다만 꼭 개선하실 필요는 없습니다.

kargowild replied:
아래와 같은 코드로 개선해보았습니다!

```java
    private Cars createCars(final List<String> carNames) {
        final List<Car> cars = carNames.stream()
                .map(Car::new)
                .collect(Collectors.toList());
        
        return new Cars(cars);
    }
```

Stream을 사용할 때와 For문을 사용할 때 내부적인 동작 차이가 있다는 말을 어디서 봤던 것 같은데, 이에 대해 추후 정리해보도록 하겠습니다!

pci2676 replied:
collect한 뒤 바로 특정 객체로 변환한다면 아래와 같이 표현할 수 있습니다.

```java
    return carNames.stream()
                .map(Car::new)
                .collect(Collectors.collectingAndThen(Collectors.toList(), Cars::new));
```

pci2676 commented on src/main/java/racingcar/domain/Referee.java:
```diff
@@ -1,15 +1,35 @@
+                .mapToInt(position -> position)
+                .max();
+
+        if (winnerPosition.isPresent()) {
+            return winnerPosition.getAsInt();
+        }
+        return 0;
```
orElse를 활용하면 간단하게 표현할 수 있을것 같습니다.

pci2676 commented on src/main/java/racingcar/CarRacing.java:
```diff
@@ -1,60 +1,60 @@
+        tryMove(tryCount, cars);
+    }
+
+    private void tryMove(final TryCount tryCount, final Cars cars) {
```
지금은 move할때마다 I/O가 이루어지고 있습니다.

이는 고비용으로 이어질 수 있는 부분일 수 있습니다.

꼭 개선할 필요는 없지만 한번 고민해보시면 좋을것 같습니다.

---

CarRacing이 controller의 역할을 한다고 말씀하셨는데요.

그렇다면 for-loop를 활용하는, 도메인 객체를 어떻게 실행하는지 구체적인 사항이 노출되어야 할 필요가 있을까요?

이 부분도 고민해보시면 좋을것 같습니다.

kargowild replied:
1. 맞습니다. 그런데 I/O를 마지막에 한 번만 호출하는 방식으로 개선해보고 싶었는데 문제가 있어요. 
예를 들어 3회 이동이 이루어졌다면, 1회 이동시의 기록, 2회 이동시의 기록, 3회 이동시의 기록을 각 자동차의 이름과 매칭하여 가지고 있어야하는데 이 부분을 어떻게 설계해야할 지 잘 떠오르질 않습니다 ㅠㅠ

2. 밑에 있는 리뷰 내용은 잘 이해가 가질 않아요. `도메인 객체를 어떻게 실행하는지 구체적인 사항이 노출된다`는 말의 의미를 다시 한 번 설명해주실 수 있을까요 ?!

pci2676 replied:
>맞습니다. 그런데 I/O를 마지막에 한 번만 호출하는 방식으로 개선해보고 싶었는데 문제가 있어요.
예를 들어 3회 이동이 이루어졌다면, 1회 이동시의 기록, 2회 이동시의 기록, 3회 이동시의 기록을 각 자동차의 이름과 매칭하여 가지고 있어야하는데 이 부분을 어떻게 설계해야할 지 잘 떠오르질 않습니다 ㅠㅠ

매 회차 car 들의 스냅샷을 찍어놓는다 생각하면 어떨까 하고 리뷰를 드렸습니다. 
회차라는 객체가 있으면 회차 안에는 각 car들의 이름과 position이 있는 map정도가 있을수 있지 않을까 싶습니다.

---
controller는 view와 domain의 상호작용을 위한 객체라고 보신것 같았습니다.  
그렇다면 controller에서 for-loop가 있는건 상호작용 자체를 나타내는 것일까요?  
제가 보기에는 상호작용을 **어떻게** 하는지에 중점이 있는 코드라고 보입니다.

그래서 controller에서 for-loop를 수행하고 있는 부분을 기존 객체에게 위임하거나 새로운 객체에 할당하는 것을 고민해보시는 것도 좋을것 같아 리뷰를 드렸습니다.

이 부분은 이후 미션을 통해 자연스럽게 터득하실것 같습니다. 지금은 고민정도만 해보셔도 충분할 것 같습니다.

kargowild replied:
1. 이 부분도 읽기 전용 모델을 적극 사용하면 도움이 될 것 같네요. 정리가 되었어요 !!
2. 그렇군요. 말씀을 듣고보니 for-loop를 돌리는 것은 컨트롤러가 맡아야 할 역할은 아닌 것 같아요. tryCount 만큼 for loop를 돌리며 `Cars.tryMove()`를 호출해줘야 하는 책임은 다른 객체가 수행해주는 것이 더 자연스러울 것 같아요. 컨트롤러는 해당 객체에게 `TryCount`, `Cars` 객체를 전달해주기만 하면 되는거구요. 생각이 많이 정리되는 것 같아요. 좋은 지적 감사드립니다!!! :)

pci2676 reviewed (changes requested):
조금 더 고민해보실 부분에 리뷰를 남겨두었으니 확인 부탁드리겠습니다.

kargowild committed:
refactor: view가 Car 객체를 받지 않고 String, int 정보를 받도록 수정

Car는 내부 값이 변경될 수 있는 메서드를 가지고 있으므로 view가 Car 객체를 직접적으로 받게되면 Car의 내부 값을 변경시킬 위험이 있으므로 리팩토링함

kargowild committed:
refactor: createCars 메서드에 stream 적용하도록 리팩토링

kargowild committed:
refactor: orElse 적용하여 코드 가독성 향상

kargowild committed:
refactor: CarAccelerator 객체는 공유해도 괜찮으므로 한 번만 생성되도록 변경

pci2676 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -45,12 +46,13 @@ public List<Car> getCars() {
 
-    public int getWinnerPosition() {
-        int maxPosition = Integer.MIN_VALUE;
+    public Map<String, Integer> getCarsPosition() {
```
👍

pci2676 reviewed (approved):
질문해주신 부분에 대해 코멘트를 남겨두었으니 확인 부탁드리겠습니다.

자동차 경주 미션 수행하시느라 고생 많으셨습니다.

머지하겠습니다. 🚀
